### ISSUE TITLE
Code error in class RCharacterList.SortDescriptors()

### ISSUE THREAD

**Issue by lijinliangyihao** [javacc/javacc#193]
I believe there is an code error in RCharacterList, method SortDescriptors, line 487. Here's my test case:
```
public class RCharacterListTest extends TestCase {

    public void testSortDescriptors1() {
        RCharacterList rlist = new RCharacterList();
        rlist.descriptors.add(new CharacterRange('e', 'g'));
        rlist.descriptors.add(new CharacterRange('a', 'd'));
        rlist.descriptors.add(new CharacterRange('h', 'k'));
        rlist.SortDescriptors();
        for (Object descriptor : rlist.descriptors) {
            CharacterRange x = (CharacterRange) descriptor;
            System.out.println(x.getLeft() + ", " + x.getRight());
        }
    }

    public void testSortDescriptors2() {
        RCharacterList rlist = new RCharacterList();
        rlist.descriptors.add(new CharacterRange('e', 'm'));
        rlist.descriptors.add(new CharacterRange('a', 'g'));
        rlist.SortDescriptors();
        for (Object descriptor : rlist.descriptors) {
            CharacterRange x = (CharacterRange) descriptor;
            System.out.println(x.getLeft() + ", " + x.getRight());
        }
    }
}
```

In ```testSortDescriptors1``` the result is
```
a, d
e, g
h, k
```
and that is correct. But in `testSortDescriptors2` the result is 
```
a, g
e, b
```
We can see that in line 2 the end of CharacterRange is smaller than the start of it. I don't think this is right.
The reason lie in the `else if` block in line 484 of class RCharacterList, which reads
```
else if (Overlaps((CharacterRange)newDesc.get(j), range))
                 {
                    CharacterRange tmp = range;
//                 we know that newDesc.get(j).left is less than or equals to range.right
//                 and here we want "range" to be set at index j, so it is newDesc.get(j).left that should be changed, other than newDesc.get(j).right
//                 so this line should be 
//                 ((CharacterRange)newDesc.get(j)).setLeft((char)(range.getRight() + 1));
                    ((CharacterRange)newDesc.get(j)).setRight((char)(range.getLeft() + 1));
                    range = (CharacterRange)newDesc.get(j);
                    newDesc.set(j, tmp);
                 }
```
After change the line to the code as I commented, the result is 
```
a, g
h, m
```
which is the result we expect.

**Comment by zosrothko** [comment]
Which version are you using? Can you upload your grammar?

**Comment by lijinliangyihao** [comment]
Master branch. I've also checked other branches, they're all the same. The suspected line is in javacc source code(not java file generated by javacc), file path is `src/main/java/org/javacc/parser/RCharacterList.java`, in method `SortDescriptors`,  line 487, would you please check out if I'm right?

**Comment by lijinliangyihao** [comment]
> Which version are you using? Can you upload your grammar?

@zosrothko

**Comment by MarcMazas** [comment]
Hi. This looks to me a real bug, detected by a junit test case. Nothing done on it?

**Comment by lijinliangyihao** [comment]
@MarcMazas No. I've already wrote how to fix it in the issue, but got no reply, so I close the issue.

**Comment by Betty-Wong** [comment]
The sortDescriptors() method seems to be implemented more cleanly here: 
https://github.com/javacc21/javacc21/blob/master/src/main/java/com/javacc/lexgen/NfaBuilder.java#L405

**Comment by zosrothko** [comment]
Fixed in javacc master branch, commit b34b17f6507f6e929bf4392a4ef3063e6ea3c8bf
To back port to JavaCC 8
